libraries/rush-lib/src/logic/operations/ShardedPhaseOperationPlugin.ts

```diff
@@ -0,0 +1,162 @@
+    hooks.createOperations.tap(PLUGIN_NAME, spliceShards);
+  }
+}
+
+function spliceShards(existingOperations: Set<Operation>, context: ICreateOperationsContext): Set<Operation> {
+  const { projectConfigurations, rushConfiguration } = context;
+
+  const shardableOperations: Set<Operation> = new Set<Operation>();
+
+  for (const operation of existingOperations) {
+    const { associatedPhase: phase, associatedProject: project } = operation;
+    const operationSettings: IOperationSettings | undefined = getOperationSettings(phase, project);
+    if (operationSettings?.sharding) {
+      shardableOperations.add(operation);
+    }
+  }
+  const getCustomParameterValuesForPhase: (phase: IPhase) => ReadonlyArray<string> =
+    getCustomParameterValuesByPhase();
+
+  for (const operation of shardableOperations) {
+    const { associatedPhase: phase, associatedProject: project } = operation;
+    const operationSettings: IOperationSettings | undefined = getOperationSettings(phase, project);
+    if (phase && project && operationSettings?.sharding && operationSettings.sharding.count > 1) {
+      existingOperations.delete(operation);
+
+      const collatorNode: Operation = new Operation({
+        phase,
+        project
+      });
+      existingOperations.add(collatorNode);
+      const collatorDisplayName: string = `${getDisplayName(phase, project)} - collate`;
+      const rawScript: string | undefined = project.packageJson.scripts?.[`${phase.name}:collate`];
+
+      if (rawScript) {
+        const collatorRunner: ShellOperationRunner = new ShellOperationRunner({
+          commandToRun: rawScript,
+          displayName: collatorDisplayName,
+          phase,
+          rushConfiguration,
+          rushProject: project,
+          operationSettings
+        });
+        collatorNode.runner = collatorRunner;
+      } else {
+        collatorNode.runner = new NullOperationRunner({
+          name: collatorDisplayName,
+          result: OperationStatus.NoOp,
+          silent: phase.missingScriptBehavior === 'silent'
+        });
+      }
+      for (const dependent of operation.consumers) {
+        dependent.addDependency(collatorNode);
+        dependent.deleteDependency(operation);
+      }
+
+      const shards: number = operationSettings.sharding.count;
+      for (const shard of Array.from({ length: shards }, (element, index) => index + 1)) {
+        let customParameters: readonly string[] = getCustomParameterValuesForPhase(phase);
+
+        // Add the shard argument to the custom parameters, replacing any templated values.
+        const shardArgumentFormat: string =
+          operationSettings.sharding.shardArgumentFormat ?? '--shard={shardIndex}/{shardCount}';
+        const outputDirectory: string = `.rush/shards/${shard}`;
+        const shardArgument: string = shardArgumentFormat
+          .replace('{shardIndex}', shard.toString())
+          .replace('{shardCount}', shards.toString());
```

Comment:
Use `replaceAll` just to be safe. Technically the argument format could place one of the values in the arguments list multiple times, even if that would be strange.

Also, these tokens should be constants, since we will want to have a step that verifies for their existence in the `shardArgumentFormat` property as part of validating the config.

Reply:
Agreed, looks like `replaceAll` is an es2021 feature, is there a rush polyfill?